#### Problem 1
- 請建立 Triangle 類別，需要給三個邊(a, b, c)的值。
- 同樣可以計算 area, perimeter, 與顯示(about)。
- 由於三角形的邊長需要滿足: 任兩邊和大於第三邊的規則。
- 所以，如果給的邊長不能構成三角形的話，就將其改成每邊為 1 的正三角形。
- 另外，area 需要使用 Heron's formula 來算。 

In [13]:
import math

class Triangle:
    def __init__(self, a, b, c):
        if(a+b>c and a+c>b and b+c>a):
            self.a = a
            self.b = b
            self.c = c
        else:
            self.a = 1
            self.b = 1
            self.c = 1

    def get_perimeter(self):
        return self.a + self.b + self.c
    
    def get_area(self):
        s = (self.a + self.b + self.c)/2
        area = math.sqrt(s*(s-self.a)*(s-self.b)*(s-self.c))
        return area
    
    def about(self):
        print(f"Side A: {self.a}")
        print(f"Side B: {self.b}")
        print(f"Side C: {self.c}")
        print(f"Area: {self.get_area()}")
        print(f"Perimeter: {self.get_perimeter()}")

A = Triangle(3, 2, 1)
A.about()

Side A: 1
Side B: 1
Side C: 1
Area: 0.4330127018922193
Perimeter: 3


#### Problem 2
- 請寫Python程式，印出你出生那個月的月曆。
- (*hint: 使用 datetime 來獲取必要的訊息。)
- 例如，以 2020 年 二月 的月曆為例，

<table>
    <tr>
        <th>Sun</th> <th>Mon</th> <th>Tue</th> <th>Wed</th> <th>Thu</th> <th>Fri</th> <th>Sat</th> 
    </tr>
    <tr>
        <th> </th> <th> </th> <th> </th> <th> </th> <th> </th> <th> </th> <th>1</th> 
    </tr>
    <tr>
        <th>2</th> <th>3</th> <th>4</th> <th>5</th> <th>6</th> <th>7</th> <th>8</th> 
    </tr>
    <tr>
        <th>9</th> <th>10</th> <th>11</th> <th>12</th> <th>13</th> <th>14</th> <th>15</th> 
    </tr>
    <tr>
        <th>16</th><th>17</th> <th>18</th> <th>19</th> <th>20</th> <th>21</th> <th>22</th>
    </tr>
    <tr>
        <th>23</th><th>24</th> <th>25</th> <th>26</th> <th>27</th> <th>28</th> <th>29</th>
    </tr>
</table>

In [48]:
import datetime as dt
year = 2001
month = 5

calendar = [["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"], 
            ["日", "一", "二", "三", "四", "五", "六"]]

def is_leap_year(year):
    if (year % 400 == 0) or ((year % 4 == 0) and (year % 100 != 0)):
        return True
    else:
        return False

def blanks(weekday):
    return (weekday + 1) % 7

def month_calendar(year:int, month:int):
    week = []
    first = dt.datetime(year, month, 1)
    if month == 1 or 3 or 5 or 7 or 8 or 10 or 12:
        days = 31
    else:
        days = 30
    blank = blanks(first.weekday())
    for i in range(blank):
        week.append(" ")
    if(month == 2):
        if(is_leap_year(year)):
            days = 29
        else:
            days = 28

# for week in calendar:
#     print(week)
# week = []


'''

monday: 0,  sunday: 6

(weekday() + 1) % 7

'''




['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
['日', '一', '二', '三', '四', '五', '六']
1


#### Problem 3
- 請讀取 pytz 中將所有時區資料(即 pytz.all_timezones)，再依各區(各區/地點)進行分類，
- 並以下列的表格來呈現，其中第一個欄位是 th tag，第二個是 td tag。
- (* 注意！有些資料並不符合(各區/地點)這種格式，可將其略去。)
<table>
    <tr>
        <th>Africa</th>
        <th> Abidjan, Accra, Addis_Ababa, Algiers, Asmara,
            Asmera, Bamako, Bangui, Banjul, Bissau,
            Blantyre, Brazzaville, Bujumbura, Cairo, Casablanca,
            Ceuta, Conakry, Dakar, Dar_es_Salaam, Djibouti,
            Douala, El_Aaiun, Freetown, Gaborone, Harare,
            Johannesburg, Juba, Kampala, Khartoum, Kigali,
            Kinshasa, Lagos, Libreville, Lome, Luanda,
            Lubumbashi, Lusaka, Malabo, Maputo, Maseru,
            Mbabane, Mogadishu, Monrovia, Nairobi, Ndjamena,
            Niamey, Nouakchott, Ouagadougou, Porto-Novo, Sao_Tome,
            Timbuktu, Tripoli, Tunis, Windhoek
        </th>
    </tr>
    <tr>
        <th>
            America
        </th>
        <th>
            Adak, Anchorage, Anguilla, Antigua, Araguaina,
Argentina, Argentina, Argentina, Argentina, Argentina,
Argentina, Argentina, Argentina, Argentina, Argentina,
Argentina, Argentina, Argentina, Aruba, Asuncion,
Atikokan, Atka, Bahia, Bahia_Banderas, Barbados,
Belem, Belize, Blanc-Sablon, Boa_Vista, Bogota,
Boise, Buenos_Aires, Cambridge_Bay, Campo_Grande, Cancun,
Caracas, Catamarca, Cayenne, Cayman, Chicago,
Chihuahua, Coral_Harbour, Cordoba, Costa_Rica, Creston,
Cuiaba, Curacao, Danmarkshavn, Dawson, Dawson_Creek,
Denver, Detroit, Dominica, Edmonton, Eirunepe,
El_Salvador, Ensenada, Fort_Nelson, Fort_Wayne, Fortaleza,
Glace_Bay, Godthab, Goose_Bay, Grand_Turk, Grenada,
Guadeloupe, Guatemala, Guayaquil, Guyana, Halifax,
Havana, Hermosillo, Indiana, Indiana, Indiana,
Indiana, Indiana, Indiana, Indiana, Indiana,
Indianapolis, Inuvik, Iqaluit, Jamaica, Jujuy,
Juneau, Kentucky, Kentucky, Knox_IN, Kralendijk,
La_Paz, Lima, Los_Angeles, Louisville, Lower_Princes,
Maceio, Managua, Manaus, Marigot, Martinique,
Matamoros, Mazatlan, Mendoza, Menominee, Merida,
Metlakatla, Mexico_City, Miquelon, Moncton, Monterrey,
Montevideo, Montreal, Montserrat, Nassau, New_York,
Nipigon, Nome, Noronha, North_Dakota, North_Dakota,
North_Dakota, Nuuk, Ojinaga, Panama, Pangnirtung,
Paramaribo, Phoenix, Port-au-Prince, Port_of_Spain, Porto_Acre,
Porto_Velho, Puerto_Rico, Punta_Arenas, Rainy_River, Rankin_Inlet,
Recife, Regina, Resolute, Rio_Branco, Rosario,
Santa_Isabel, Santarem, Santiago, Santo_Domingo, Sao_Paulo,
Scoresbysund, Shiprock, Sitka, St_Barthelemy, St_Johns,
St_Kitts, St_Lucia, St_Thomas, St_Vincent, Swift_Current,
Tegucigalpa, Thule, Thunder_Bay, Tijuana, Toronto,
Tortola, Vancouver, Virgin, Whitehorse, Winnipeg,
Yakutat, Yellowknife
        </th>
    </tr>
</table>

In [20]:
import pytz

timezones = {"Africa":[], "America":[], "Antarctica":[], "Arctic":[], "Asia":[], "Atlantic":[], 
             "Australia":[], "Brazil":[], "Canada":[], "Chile":[], "Europe":[], "Indian":[], 
             "Mexico":[], "Pacific":[], "US":[]}

for timezone in pytz.all_timezones:
    countries = timezone.split("/")
    if(timezones.get(countries[0])!= None):
        timezones[countries[0]].append(countries[1])

table = "<table>\n"

for timezone in timezones.items():
    table += "\t<tr>\n"
    table += f"\t\t<th>{timezone[0]}</th>\n"
    table += f"\t\t<th>"
    for region in timezone[1]:
        table += f"{region}, "
    table = table[:-2]
    table += "</th>\n"
    table += "\t</tr>\n"

table += "</table>"

print(table)



<table>
	<tr>
		<th>Africa</th>
		<th>Abidjan, Accra, Addis_Ababa, Algiers, Asmara, Asmera, Bamako, Bangui, Banjul, Bissau, Blantyre, Brazzaville, Bujumbura, Cairo, Casablanca, Ceuta, Conakry, Dakar, Dar_es_Salaam, Djibouti, Douala, El_Aaiun, Freetown, Gaborone, Harare, Johannesburg, Juba, Kampala, Khartoum, Kigali, Kinshasa, Lagos, Libreville, Lome, Luanda, Lubumbashi, Lusaka, Malabo, Maputo, Maseru, Mbabane, Mogadishu, Monrovia, Nairobi, Ndjamena, Niamey, Nouakchott, Ouagadougou, Porto-Novo, Sao_Tome, Timbuktu, Tripoli, Tunis, Windhoek</th>
	</tr>
	<tr>
		<th>America</th>
		<th>Adak, Anchorage, Anguilla, Antigua, Araguaina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Aruba, Asuncion, Atikokan, Atka, Bahia, Bahia_Banderas, Barbados, Belem, Belize, Blanc-Sablon, Boa_Vista, Bogota, Boise, Buenos_Aires, Cambridge_Bay, Campo_Grande, Cancun, Caracas, Catamarca, Cayenne, Cayman, Chicago, Chihu

<table>
	<tr>
		<th>Africa</th>
		<th>Abidjan, Accra, Addis_Ababa, Algiers, Asmara, Asmera, Bamako, Bangui, Banjul, Bissau, Blantyre, Brazzaville, Bujumbura, Cairo, Casablanca, Ceuta, Conakry, Dakar, Dar_es_Salaam, Djibouti, Douala, El_Aaiun, Freetown, Gaborone, Harare, Johannesburg, Juba, Kampala, Khartoum, Kigali, Kinshasa, Lagos, Libreville, Lome, Luanda, Lubumbashi, Lusaka, Malabo, Maputo, Maseru, Mbabane, Mogadishu, Monrovia, Nairobi, Ndjamena, Niamey, Nouakchott, Ouagadougou, Porto-Novo, Sao_Tome, Timbuktu, Tripoli, Tunis, Windhoek</th>
	</tr>
	<tr>
		<th>America</th>
		<th>Adak, Anchorage, Anguilla, Antigua, Araguaina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Argentina, Aruba, Asuncion, Atikokan, Atka, Bahia, Bahia_Banderas, Barbados, Belem, Belize, Blanc-Sablon, Boa_Vista, Bogota, Boise, Buenos_Aires, Cambridge_Bay, Campo_Grande, Cancun, Caracas, Catamarca, Cayenne, Cayman, Chicago, Chihuahua, Ciudad_Juarez, Coral_Harbour, Cordoba, Costa_Rica, Creston, Cuiaba, Curacao, Danmarkshavn, Dawson, Dawson_Creek, Denver, Detroit, Dominica, Edmonton, Eirunepe, El_Salvador, Ensenada, Fort_Nelson, Fort_Wayne, Fortaleza, Glace_Bay, Godthab, Goose_Bay, Grand_Turk, Grenada, Guadeloupe, Guatemala, Guayaquil, Guyana, Halifax, Havana, Hermosillo, Indiana, Indiana, Indiana, Indiana, Indiana, Indiana, Indiana, Indiana, Indianapolis, Inuvik, Iqaluit, Jamaica, Jujuy, Juneau, Kentucky, Kentucky, Knox_IN, Kralendijk, La_Paz, Lima, Los_Angeles, Louisville, Lower_Princes, Maceio, Managua, Manaus, Marigot, Martinique, Matamoros, Mazatlan, Mendoza, Menominee, Merida, Metlakatla, Mexico_City, Miquelon, Moncton, Monterrey, Montevideo, Montreal, Montserrat, Nassau, New_York, Nipigon, Nome, Noronha, North_Dakota, North_Dakota, North_Dakota, Nuuk, Ojinaga, Panama, Pangnirtung, Paramaribo, Phoenix, Port-au-Prince, Port_of_Spain, Porto_Acre, Porto_Velho, Puerto_Rico, Punta_Arenas, Rainy_River, Rankin_Inlet, Recife, Regina, Resolute, Rio_Branco, Rosario, Santa_Isabel, Santarem, Santiago, Santo_Domingo, Sao_Paulo, Scoresbysund, Shiprock, Sitka, St_Barthelemy, St_Johns, St_Kitts, St_Lucia, St_Thomas, St_Vincent, Swift_Current, Tegucigalpa, Thule, Thunder_Bay, Tijuana, Toronto, Tortola, Vancouver, Virgin, Whitehorse, Winnipeg, Yakutat, Yellowknife</th>
	</tr>
	<tr>
		<th>Antarctica</th>
		<th>Casey, Davis, DumontDUrville, Macquarie, Mawson, McMurdo, Palmer, Rothera, South_Pole, Syowa, Troll, Vostok</th>
	</tr>
	<tr>
		<th>Arctic</th>
		<th>Longyearbyen</th>
	</tr>
	<tr>
		<th>Asia</th>
		<th>Aden, Almaty, Amman, Anadyr, Aqtau, Aqtobe, Ashgabat, Ashkhabad, Atyrau, Baghdad, Bahrain, Baku, Bangkok, Barnaul, Beirut, Bishkek, Brunei, Calcutta, Chita, Choibalsan, Chongqing, Chungking, Colombo, Dacca, Damascus, Dhaka, Dili, Dubai, Dushanbe, Famagusta, Gaza, Harbin, Hebron, Ho_Chi_Minh, Hong_Kong, Hovd, Irkutsk, Istanbul, Jakarta, Jayapura, Jerusalem, Kabul, Kamchatka, Karachi, Kashgar, Kathmandu, Katmandu, Khandyga, Kolkata, Krasnoyarsk, Kuala_Lumpur, Kuching, Kuwait, Macao, Macau, Magadan, Makassar, Manila, Muscat, Nicosia, Novokuznetsk, Novosibirsk, Omsk, Oral, Phnom_Penh, Pontianak, Pyongyang, Qatar, Qostanay, Qyzylorda, Rangoon, Riyadh, Saigon, Sakhalin, Samarkand, Seoul, Shanghai, Singapore, Srednekolymsk, Taipei, Tashkent, Tbilisi, Tehran, Tel_Aviv, Thimbu, Thimphu, Tokyo, Tomsk, Ujung_Pandang, Ulaanbaatar, Ulan_Bator, Urumqi, Ust-Nera, Vientiane, Vladivostok, Yakutsk, Yangon, Yekaterinburg, Yerevan</th>
	</tr>
	<tr>
		<th>Atlantic</th>
		<th>Azores, Bermuda, Canary, Cape_Verde, Faeroe, Faroe, Jan_Mayen, Madeira, Reykjavik, South_Georgia, St_Helena, Stanley</th>
	</tr>
	<tr>
		<th>Australia</th>
		<th>ACT, Adelaide, Brisbane, Broken_Hill, Canberra, Currie, Darwin, Eucla, Hobart, LHI, Lindeman, Lord_Howe, Melbourne, NSW, North, Perth, Queensland, South, Sydney, Tasmania, Victoria, West, Yancowinna</th>
	</tr>
	<tr>
		<th>Brazil</th>
		<th>Acre, DeNoronha, East, West</th>
	</tr>
	<tr>
		<th>Canada</th>
		<th>Atlantic, Central, Eastern, Mountain, Newfoundland, Pacific, Saskatchewan, Yukon</th>
	</tr>
	<tr>
		<th>Chile</th>
		<th>Continental, EasterIsland</th>
	</tr>
	<tr>
		<th>Europe</th>
		<th>Amsterdam, Andorra, Astrakhan, Athens, Belfast, Belgrade, Berlin, Bratislava, Brussels, Bucharest, Budapest, Busingen, Chisinau, Copenhagen, Dublin, Gibraltar, Guernsey, Helsinki, Isle_of_Man, Istanbul, Jersey, Kaliningrad, Kiev, Kirov, Kyiv, Lisbon, Ljubljana, London, Luxembourg, Madrid, Malta, Mariehamn, Minsk, Monaco, Moscow, Nicosia, Oslo, Paris, Podgorica, Prague, Riga, Rome, Samara, San_Marino, Sarajevo, Saratov, Simferopol, Skopje, Sofia, Stockholm, Tallinn, Tirane, Tiraspol, Ulyanovsk, Uzhgorod, Vaduz, Vatican, Vienna, Vilnius, Volgograd, Warsaw, Zagreb, Zaporozhye, Zurich</th>
	</tr>
	<tr>
		<th>Indian</th>
		<th>Antananarivo, Chagos, Christmas, Cocos, Comoro, Kerguelen, Mahe, Maldives, Mauritius, Mayotte, Reunion</th>
	</tr>
	<tr>
		<th>Mexico</th>
		<th>BajaNorte, BajaSur, General</th>
	</tr>
	<tr>
		<th>Pacific</th>
		<th>Apia, Auckland, Bougainville, Chatham, Chuuk, Easter, Efate, Enderbury, Fakaofo, Fiji, Funafuti, Galapagos, Gambier, Guadalcanal, Guam, Honolulu, Johnston, Kanton, Kiritimati, Kosrae, Kwajalein, Majuro, Marquesas, Midway, Nauru, Niue, Norfolk, Noumea, Pago_Pago, Palau, Pitcairn, Pohnpei, Ponape, Port_Moresby, Rarotonga, Saipan, Samoa, Tahiti, Tarawa, Tongatapu, Truk, Wake, Wallis, Yap</th>
	</tr>
	<tr>
		<th>US</th>
		<th>Alaska, Aleutian, Arizona, Central, East-Indiana, Eastern, Hawaii, Indiana-Starke, Michigan, Mountain, Pacific, Samoa</th>
	</tr>
</table>


#### Problem 4
- 讀入資料檔 BigData-Seat-2023-0302.txt，並完成下列的改寫:
- 姓名的部分: 是 2 個字的，將單名改成 *； 是 3 個字的中間那個字改成 *。
- 最後再將改寫過的結果，存入檔案 BigData-Seats.txt (encoding='utf-8')

In [42]:
fname = './BigData-Seat-2023-0302.txt'
with open(fname, "r", encoding='utf-8') as fp:
    lines = fp.read()
    lines = lines.split('\n')

result = []
for line in lines:  # line = '108590099,1,7,王*同'
    items = line.split(',')
    #items[0] = items[0][:-2] + "**"
    if len(items[3]) > 2:
        items[3] = items[3][0] + '*' + items[3][2:]
    else:
        items[3] = items[3][0] + '*'
    
    line = ','.join(items)    
    result.append(line)
    
print(result)
with open('BigData-Seat.txt', "w", encoding='utf-8') as fp:
    fp.writelines('\n'.join(result))

['108590020,1,7,林*凱', '110590053,3,4,許*誠', '110590069,2,7,黃*義', '108310217,4,-4,蔡*緯', '108AB0047,3,7,林*妍', '109310323,2,-4,陳*諭', '110590054,3,5,梁*軒', '110590044,2,6,黃*輝', '108AB0025,3,6,陳*渝', '110590452,2,-1,莊*潔', '108450021,1,9,賴*勳', '108ac1026,2,4,林*', '109AB0051,1,6,陳*翔', '1082B0032,4,7,洪*瑋', '108310227,4,-5,劉*碩', '110590055,3,3,林*', '108590009,1,10,蔡*宏', '110590064,4,8,劉*軒', '108ab0014,3,14,彭*俊', '110AC1005,3,-6,陳*晟', '108AB0701,1,-2,蘇*覡', '110590048,4,3,陳*諺', '110590004,1,2,林*廷', '107590027,4,-1,許*瑋', '108AB0744,1,12,廖*其', '110810006,3,-7,林*曆', '110590041,1,-1,陳*堯', '110590017,4,-14,陳*安', '1082B0005,4,5,林*伶', '110590007,4,4,陸*寰', '110590065,4,9,陳*翰', '110590028,4,2,黃*鈞', '109820004,2,2,陳*愷', '110590066,1,3,吳*駒', '110590058,1,4,林*憲', '1082B0009,4,6,李*維', '110590045,2,-4,陸*康', '109310207,2,-5,鄭*恩', '108820018,4,-2,蔡*宇']


#### Problem 5
- 承上一題，將 BigData-Seats.txt 中，資料標記的位置，把所有人名與學號，填入 HTML 表格對應的位置。
- HTML 表格寫作參考:
<table>
    <tr>
        <th>(1, 1)</th> <th>(1, 2)</th> <th>(1, 3)</th>
    </tr>
    <tr>
        <th>(2, 1)</th> <th>(2, 2)</th> <th>(2, 3)</th>
    </tr>
</table>

In [43]:
fname = './BigData-Seat.txt'
seat = [[] for i in range(4)]
result = "<table>\n"

for row in seat:
    for i in range(14):
        row.append(0)

with open(fname, "r", encoding='utf-8') as fp:
    lines = fp.read()
    lines = lines.split('\n')

for line in lines:
    # line = '108590099,1,7,王*同'
    items = line.split(',')
    row = int(items[1])
    col = int(items[2])
    # print(items[2])
    if(col < 0):
        seat[row-1][14+col] = f"{items[0]}<br>{items[3]}"
    else:
        seat[row-1][col-1] = f"{items[0]}<br>{items[3]}"

for row in seat:
    result += "\t<tr>\n"
    for col in row:
        result += f"\t\t<th>{col}</th>\n"
    result += "\t</tr>\n"
result += "</table>"

with open('BigData-Seat.html', "w", encoding='utf-8') as fp:
    fp.writelines(result)



<table>
	<tr>
		<th>0</th>
		<th>110590004<br>林*廷</th>
		<th>110590066<br>吳*駒</th>
		<th>110590058<br>林*憲</th>
		<th>0</th>
		<th>109AB0051<br>陳*翔</th>
		<th>108590020<br>林*凱</th>
		<th>0</th>
		<th>108450021<br>賴*勳</th>
		<th>108590009<br>蔡*宏</th>
		<th>0</th>
		<th>108AB0744<br>廖*其</th>
		<th>108AB0701<br>蘇*覡</th>
		<th>110590041<br>陳*堯</th>
	</tr>
	<tr>
		<th>0</th>
		<th>109820004<br>陳*愷</th>
		<th>0</th>
		<th>108ac1026<br>林寧</th>
		<th>0</th>
		<th>110590044<br>黃*輝</th>
		<th>110590069<br>黃*義</th>
		<th>0</th>
		<th>0</th>
		<th>109310207<br>鄭*恩</th>
		<th>110590045<br>陸*康</th>
		<th>0</th>
		<th>0</th>
		<th>110590452<br>莊*潔</th>
	</tr>
	<tr>
		<th>0</th>
		<th>0</th>
		<th>110590055<br>林蔚</th>
		<th>110590053<br>許*誠</th>
		<th>110590054<br>梁*軒</th>
		<th>108AB0025<br>陳*渝</th>
		<th>108AB0047<br>林*妍</th>
		<th>110810006<br>林*曆</th>
		<th>110AC1005<br>陳*晟</th>
		<th>0</th>
		<th>0</th>
		<th>0</th>
		<th>0</th>
		<th>108ab0014<br>彭*俊</th>
	</tr>
	<tr>
		<th>110590017<br>陳*安</th>
		<th>110590028<br>黃*鈞</th>
		<th>110590048<br>陳*諺</th>
		<th>110590007<br>陸*寰</th>
		<th>1082B0005<br>林*伶</th>
		<th>1082B0009<br>李*維</th>
		<th>1082B0032<br>洪*瑋</th>
		<th>110590064<br>劉*軒</th>
		<th>110590065<br>陳*翰</th>
		<th>108310227<br>劉*碩</th>
		<th>108310217<br>蔡*緯</th>
		<th>0</th>
		<th>108820018<br>蔡*宇</th>
		<th>107590027<br>許*瑋</th>
	</tr>
</table>
